### PR TITLE
python3Packages.protonup: init at 0.1.4

### PR DESCRIPTION
**Pull request by flexagoon**
###### Motivation for this change
Protonup is a cli utility to easily install and update GloriousEgroll's fork of Proton
https://github.com/AUNaseef/protonup

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 135948` at 0c2a4daa run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>protonup (python39Packages.protonup)</li>
    <li>python38Packages.protonup</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/python-include-tests.md">python-include-tests</a>

Consider adding a `checkPhase` for tests, or if not feasible, `pythonImportsCheck`.

Near pkgs/tools/package-management/protonup/default.nix:23:0:
```
   |
23 |     description = "CLI program and API to automate the installation and update of GloriousEggroll's Proton-GE";
   | ^
```
</li>
  </ul>
</details>

File: pkgs/tools/package-management/protonup/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  };
+
+  postPatch = ''
+    substituteInPlace setup.cfg \
+      --replace "argparse" ""
+  '';
+
+  propagatedBuildInputs = [ requests configparser ];
+
+  doCheck = false;
```
Comment (fabaff) [inline pkgs/tools/package-management/protonup/default.nix:19]: Please add a comment why the tests are disabled.

File: pkgs/tools/package-management/protonup/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
```
Comment (fabaff) [inline pkgs/tools/package-management/protonup/default.nix:20]: Please add `pythonImportsCheck` as a safe-guard.

File: pkgs/tools/package-management/protonup/default.nix [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  version = "0.1.4";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0z5q0s9h51w2bqm9lkafml14g13v2dgm4nm9x06v7nxqc9msmyyy";
+  };
+
+  postPatch = ''
+    substituteInPlace setup.cfg \
+      --replace "argparse" ""
```
Comment (fabaff) [inline pkgs/tools/package-management/protonup/default.nix:14]: I proposed a PR to get it fixed upstream.

Reference: https://github.com/AUNaseef/protonup/pull/14

**Review by fabaff — COMMENTED** [review]
Could you please update the commit message?

`python3Packages.protonup: init at 0.1.4`

File: pkgs/tools/package-management/protonup/default.nix [hunk 4]
```diff
@@ -0,0 +1,27 @@
+{ lib, buildPythonPackage, fetchPypi, requests, configparser }:
+
+buildPythonPackage rec {
+  pname = "protonup";
+  version = "0.1.4";
+
```
Comment (fabaff) [inline pkgs/tools/package-management/protonup/default.nix:6]: Just to be on the save site.

```suggestion

  disabled = pythonOlder "3.6";
```

see https://github.com/AUNaseef/protonup/blob/main/setup.cfg#L19

File: pkgs/tools/package-management/protonup/default.nix (hunk 1 above)
Comment (flexagoon) [reply pkgs/tools/package-management/protonup/default.nix:19]: Actually I don't know why I disabled tests, I will enable them.

File: pkgs/tools/package-management/protonup/default.nix (hunk 4 above)
Comment (flexagoon) [reply pkgs/tools/package-management/protonup/default.nix:6]: error: undefined variable 'pythonOlder'

Comment (flexagoon) [reply pkgs/tools/package-management/protonup/default.nix:6]: Oh, never mind. I forgot to specify it in the input parameters.

File: pkgs/tools/package-management/protonup/default.nix (hunk 1 above)
Comment (flexagoon) [reply pkgs/tools/package-management/protonup/default.nix:19]: Turns out protonup doesn't actually have any tests, I'll just use pythonImportsCheck.

File: pkgs/tools/package-management/protonup/default.nix [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{ lib, buildPythonPackage, pythonOlder, fetchPypi, requests, configparser, pytest }:
```
Comment (fabaff) [inline pkgs/tools/package-management/protonup/default.nix:1]:
```suggestion
{ lib, buildPythonPackage, pythonOlder, fetchPypi, requests, configparser }:
```

File: pkgs/tools/package-management/protonup/default.nix [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  postPatch = ''
+    substituteInPlace setup.cfg \
+      --replace "argparse" ""
+  '';
+
+  propagatedBuildInputs = [ requests configparser ];
+
+  doCheck = false; # protonup does not have any tests
+  pythonImportsCheck = [ "requests" "argparse" "configparser" ];
```
Comment (fabaff) [inline pkgs/tools/package-management/protonup/default.nix:21]: `pythonImportsCheck` for other modules should be handled by themself.

```suggestion
  pythonImportsCheck = [ "protonup" ];
```

File: pkgs/top-level/python-packages.nix [hunk 7]
```diff
@@ -5696,6 +5696,8 @@ in {
 
   protobuf3-to-dict = callPackage ../development/python-modules/protobuf3-to-dict { };
 
+  protonup = callPackage ../tools/package-management/protonup { };
```
Comment (fabaff) [inline pkgs/top-level/python-packages.nix:5699]:
```suggestion
  protonup = callPackage ../development/python-modules/protonup { };
```

**Review by fabaff — COMMENTED** [review]
Python modules have their home in `pkgs/development/python-modules/`

Could you please squash the commits for the package (keep the one for `maintainer-list.nix`)?

- `maintainers/maintainer-list.nix` -> `maintainers: add flexagoon`
- `pkgs/development/python-modules/protonu/defult.nix` - > `python3Packages.protonup: init at 0.1.4`

File: pkgs/tools/package-management/protonup/default.nix (hunk 5 above)
Comment (fabaff) [reply pkgs/tools/package-management/protonup/default.nix:1]:
```bash
nix run -f https://github.com/jtojnar/nixpkgs-hammering/archive/master.tar.gz -c nixpkgs-hammer [pkg_name]
```

can help to identify issue. This is what the bot [did](https://github.com/NixOS/nixpkgs/pull/135948#issuecomment-907534081).

**Comment by flexagoon** [comment]
@fabaff I don't think the package should be located in development/python-modules, as it's mostly used as a standalone app, and not as a library. For example, youtube-dl is located in tools/mics, although it's also a python package which сan be used as an importable library.

**Comment by fabaff** [comment]
`protonup = with python3Packages; toPythonApplication protonup;` takes care that's exposed as CLI tool. 

If you want it "as tool only" then you should use `python3.pkgs.buildPythonApplication rec {` (see "[15.21.2.3.2. buildPythonApplication function](https://nixos.org/manual/nixpkgs/stable/#reference)".

**Comment by flexagoon** [comment]
@fabaff then why is youtube-dl packaged like that? Is that wrong?

**Comment by fabaff** [comment]
> @fabaff then why is youtube-dl packaged like that? 

Because the `youtube-dl` derivation is almost 10 years old and over time only minimal adjustment were made to `youtube-dl`. `youtube-dl` is still at the original location as it was when it was introduced. Nowadays, it wouldn't be placed there or it would be using `buildPythonApplication`. 

> Is that wrong?

I wouldn't say it's wrong. It simply doesn't fit the current "common practices and standards".

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 135948` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>protonup (python39Packages.protonup)</li>
    <li>python38Packages.protonup</li>
  </ul>
</details>

**Review by fabaff — APPROVED** [review]
- [x] package path fits guidelines
- [x] package name fits guidelines
- [x] package version fits guidelines
- [x] `meta.description` is set and fits guidelines
- [x] `meta.license` fits upstream license
- [x] `meta.maintainers` is set
- [x] package builds (see https://github.com/NixOS/nixpkgs/pull/135948#issuecomment-912584749)
- [ ] all depending packages build
- [x] tested on x86_64

```bash
...ew/pr-135948]$ ./results/protonup/bin/protonup 
Ready to download Proton-6.16-GE-1 
Size      : 411.7 MiB 
Published : 2021-08-29
Continue? (Y/n): n
```

Change looks good to me.

**Pull request merged**